File: LeetCode_Marathons/LCM1/Hard/Reducing_Dishes/a.go
var globalSatisfaction []int

func getSum(count int) int {
    n := len(globalSatisfaction)
    ans, counter := 0, 1

    for i := n - count; i <= n - 1; i++ {
        ans += counter * globalSatisfaction[i];
        counter++;
    }

    return ans;
}

func maxSatisfaction(satisfaction []int) int {
    n := len(satisfaction)
    sort.Ints(satisfaction)
    globalSatisfaction = satisfaction

    l, r, mid := 0, n, 0
    for l < r {
        mid = (l + r + 1) / 2
        if getSum(mid) > getSum(mid - 1) {
            l = mid
        } else {
            r = mid - 1
        }
    }

    return getSum(l)
}

File: LeetCode_Marathons/LCM1/Hard/Reducing_Dishes/a.java
class Solution {
    int[] globalSatisfaction;

    private int getSum(int count) {
        int n = globalSatisfaction.length;
        int ans = 0, counter = 1;
        for (int i = n - count; i <= n - 1; i++) {
            ans += counter * globalSatisfaction[i];
            counter++;
        }

        return ans;
    }

    public int maxSatisfaction(int[] satisfaction) {
        int n = satisfaction.length;
        Arrays.sort(satisfaction);
        globalSatisfaction = satisfaction;
        
        int l = 0, r = n, mid = 0;
        while (l < r) {
            mid = (l + r + 1) / 2;
            if (getSum(mid) > getSum(mid - 1)) l = mid;
            else r = mid - 1;
        }

        return getSum(l);
    }
}